fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/codec/LuceneOptimizedCompoundReader.java

```diff
@@ -0,0 +1,218 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.lucene.codec;
+
+
+import org.apache.lucene.codecs.CodecUtil;
+import org.apache.lucene.index.CorruptIndexException;
+import org.apache.lucene.index.IndexFileNames;
+import org.apache.lucene.index.SegmentInfo;
+import org.apache.lucene.store.ChecksumIndexInput;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.apache.lucene.store.IndexInput;
+import org.apache.lucene.store.IndexOutput;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.util.IOUtils;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ *
+ * Borrowed from Lucene to remove checksum from end of file.
+ *
+ * Class for accessing a compound stream.
+ * This class implements a directory, but is limited to only read operations.
+ * Directory methods that would normally modify data throw an exception.
+ * @lucene.experimental
+ */
+final class LuceneOptimizedCompoundReader extends Directory {
+
+    private final Directory directory;
+    private final String segmentName;
+    private final Map<String, FileEntry> entries;
+    private final IndexInput handle;
+
+    /** Offset/Length for a slice inside of a compound file. */
+    public static final class FileEntry {
+        long offset;
+        long length;
+    }
+
+    /**
+     * Create a new CompoundFileDirectory.
+     */
+    // TODO: we should just pre-strip "entries" and append segment name up-front like simpletext?
+    // this need not be a "general purpose" directory anymore (it only writes index files)
+    public LuceneOptimizedCompoundReader(Directory directory, SegmentInfo si, IOContext context) throws IOException {
+        this.directory = directory;
+        this.segmentName = si.name;
+        String entriesFileName = IndexFileNames.segmentFileName(segmentName, "", LuceneOptimizedCompoundFormat.ENTRIES_EXTENSION);
+        this.entries = readEntries(si.getId(), directory, entriesFileName);
+        String dataFileName = IndexFileNames.segmentFileName(segmentName, "", LuceneOptimizedCompoundFormat.DATA_EXTENSION);
+        handle = directory.openInput(dataFileName, context);
+        /* REMOVED TO NOT READ THE CHECKSUM ON EACH SEARCH
+        try {
+            CodecUtil.checkIndexHeader(handle, LuceneOptimizedCompoundFormat.DATA_CODEC, version, version, si.getId(), "");
```

Comment:
this is one we can skip because it's not necessary for the implementation of the reader. I think that the point is to see if this is significantly expensive enough to make  adifference.

Reply:
umm so we never need to maintain a ```version``` here? @sophiabuell